controlpanel/kubeapi/oidc_patch.py

```diff
@@ -0,0 +1,121 @@
+import base64
+import datetime
+from datetime import timezone
+import json
+import logging
+import os
+import tempfile
+
+import kubernetes
+from kubernetes.config.kube_config import _is_expired
+import oauthlib.oauth2
+from requests_oauthlib import OAuth2Session
+import yaml
+
+
+log = logging.getLogger(__name__)
+
+
+def load_token(self, provider):
+    if "config" not in provider:
+        return
+
+    parts = provider["config"]["id-token"].split(".")
+
+    if len(parts) != 3:  # Not a valid JWT
+        return None
+
+    padding = (4 - len(parts[1]) % 4) * "="
+
+    jwt_attributes = json.loads(base64.b64decode(parts[1] + padding).decode("utf-8"))
+
+    expire = jwt_attributes.get("exp")
+
+    if (expire is not None) and (
+        _is_expired(datetime.datetime.fromtimestamp(expire, tz=timezone.utc))
+    ):
+        log.debug(provider["config"]["idp-issuer-url"])
+
+        self._refresh_oidc(provider)
+
+        if self._config_persister:
+            self._config_persister(self._config.value)
+
+    self.token = f"Bearer {provider['config']['id-token']}"
+
+    return self.token
+
+
+kubernetes.config.kube_config.KubeConfigLoader._load_oid_token = load_token
+
+
+def refresh_oidc(self, provider):
+    config = kubernetes.client.Configuration()
+
+    if 'idp-certificate-authority-data' in provider['config']:
+        ca_cert = tempfile.NamedTemporaryFile(delete=True)
+
+        ca_data = provider['config']['idp-certificate-authority-data']
+
+        padding = (4 - len(ca_data) % 4) * "="
+
+        cert = base64.b64decode(ca_data + padding).decode('utf-8')
+
+        with open(ca_cert.name, 'w') as fh:
+            fh.write(cert)
+
+        config.ssl_ca_cert = ca_cert.name
+
+    else:
+        config.verify_ssl = False
+
+    client = kubernetes.client.ApiClient(configuration=config)
+
+    response = client.request(
+        method="GET",
+        url="%s/.well-known/openid-configuration"
+        % provider['config']['idp-issuer-url'].rstrip('/'),
+    )
```

Comment:
#### 2a. Monkey-patching the Kubernetes client
>Lines 74 - 78

I submitted this patch to the Kubernetes Python client project - we get a URL ending with a `/` from Auth0 for `idp-issuer-url` (the tenant base URL). But this code assumes there is no trailing slash and ends up creating a URL that is rejected by the Kubernetes API. 🙄 

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-8c49a844762f3e9fdeb511f2a34ca23eR113)